server/api/public.py

```diff
@@ -42,15 +47,106 @@
-    sample_sizes = {
-        "ballotComparison": {},
-        "ballotPolling": {},
-        "batchComparison": {},
+    contest_for_sampler = parse_compute_sample_sizes_input(request.get_json())
+
+    audit_types = ["ballotComparison", "ballotPolling", "batchComparison"]
+    sample_sizes = {audit_type: {} for audit_type in audit_types}
+    sample_size_functions = {
+        "ballotComparison": compute_ballot_comparison_sample_size,
+        "ballotPolling": compute_ballot_polling_sample_size,
+        "batchComparison": compute_batch_comparison_sample_size,
     }
-    for risk_limit_percentage in list(range(0, 21)):
-        sample_sizes["ballotComparison"][str(risk_limit_percentage)] = 0
-        sample_sizes["ballotPolling"][str(risk_limit_percentage)] = 0
-        sample_sizes["batchComparison"][str(risk_limit_percentage)] = 0
+    for audit_type in audit_types:
+        for risk_limit_percentage in list(range(0, 21)):
+            sample_size_function = sample_size_functions[audit_type]
+            sample_sizes[audit_type][str(risk_limit_percentage)] = sample_size_function(
+                contest_for_sampler, risk_limit_percentage
+            )
+
     return jsonify(sample_sizes)
+
+
+def parse_compute_sample_sizes_input(
+    compute_sample_sizes_input: Any,
+) -> sampler_contest.Contest:
+    validate(compute_sample_sizes_input, COMPUTE_SAMPLE_SIZES_INPUT_SCHEMA)
+
+    election_results = compute_sample_sizes_input["electionResults"]
+    candidates = election_results["candidates"]
+    num_winners = election_results["numWinners"]
+    total_ballots_cast = election_results["totalBallotsCast"]
+
+    if all(candidate["votes"] == 0 for candidate in candidates):
+        raise Conflict("At least 1 candidate must have greater than 0 votes")
+
+    if num_winners >= len(candidates):
+        raise Conflict("Number of winners must be less than number of candidates")
+
+    candidate_votes_sum = sum(candidate["votes"] for candidate in candidates)
+    if total_ballots_cast < candidate_votes_sum:
+        raise Conflict("Total ballots cast cannot be less than sum of candidate votes")
+
+    contest = Contest(
+        choices=[
+            ContestChoice(
+                id=f"candidate-{i}",
+                num_votes=candidate["votes"]
+                + 0.01  # Safeguard against math errors involving 0s
+                if candidate["votes"] == 0
+                else candidate["votes"],
+            )
+            for i, candidate in enumerate(candidates)
+        ],
+        num_winners=num_winners,
+        total_ballots_cast=total_ballots_cast,
+        votes_allowed=num_winners,
+    )
+    contest_for_sampler: sampler_contest.Contest = sampler_contest.from_db_contest(
+        contest
+    )
+    return contest_for_sampler
+
+
+def is_full_hand_tally_required(
+    contest: sampler_contest.Contest, risk_limit_percentage: int
+) -> bool:
+    is_tie = contest.diluted_margin == 0
+    return is_tie or risk_limit_percentage == 0
```

Comment:
Both these situations (specifically `risk_limit_percentage == 0`) can trigger zero-related errors (e.g. ZeroDivisionErrors), so I special-case and short-circuit

Reply:
Could you add a comment explaining that?

Reply:
I think the audit math should handle this (same argument as above - this could impact real audits)

Reply:
Good point! The ballot polling and ballot comparison audit math was actually already handling the tie case. I pushed the logic for the zero risk limit, though, into those functions. (Makes sense that they weren't covering those since you can't select a risk limit of 0 in our standard UI.)

https://github.com/votingworks/arlo/pull/1600/commits/a741eb5c784f3662e27d790a718abb3c3a0bcbf5